lib/controlFlow.ts

```diff
@@ -0,0 +1,58 @@
+import { Logger } from './logger';
+import { EventEmitter } from 'events';
+import * as Q from 'q';
+let logger = new Logger('Control-Flow');
+export class ControlFlow extends EventEmitter {
+    private commandChain = Q.when(undefined);
+    private numTasks: number = 0;
+    private _isIdle: boolean = false;
+    constructor() {
+        super();
+     }
+    
+    execute(fn: () => PromiseLike<any> | any, text? : string): PromiseLike<any> {
+        this.numTasks++;
+        this._isIdle = false;
+        return this.commandChain = this.commandChain
+            .then(fn)
+            .then((value) => { 
+                if (--this.numTasks === 0) {
+                    this.emit('idle');
+                    this._isIdle = true;
+                }
+                return undefined;
```

Comment:
Command chain consists of bound functions with optional params in some cases. The returned value is of no use and might pollute the next chained command.